File: ling_chat/core/ai_service/script_engine/events/ai_dialogue_event.py
from ling_chat.core.ai_service.script_engine.events.base_event import BaseEvent
from ling_chat.core.ai_service.script_engine.utils.script_function import ScriptFunction
from ling_chat.core.logger import logger
from ling_chat.core.schemas.response_models import ResponseFactory
from ling_chat.core.service_manager import service_manager
from ling_chat.core.messaging.broker import message_broker
from ling_chat.game_database.models import LineAttribute, LineBase


class AIDialogueEvent(BaseEvent):
    """处理AI对话事件"""

    async def _execute(self):
        character = self.event_data.get('character', '')
        prompt = self.event_data.get('prompt', '')

        role = ScriptFunction.get_role(self.game_status, self.script_status, character)
        self.game_status.current_character = role

        if prompt and prompt != '':
            system_input = LineBase(content=ScriptFunction.user_message_builder("", prompt),attribute=LineAttribute.USER,display_name=self.game_status.player.user_name)
            self.game_status.add_line(system_input)

        ai_service = service_manager.ai_service
        if not ai_service:
            logger.error("AI service not found")
            return

        # TODO：获取上一个事件是否为玩家输入再执行这个，会比较优雅
        # await message_broker.publish(self.client_id, (ResponseFactory.create_thinking(True).model_dump()))
        async for response in ai_service.message_generator.process_message_stream():
            await message_broker.publish(self.client_id, response.model_dump())


    @classmethod
    def can_handle(cls, event_type: str) -> bool:
        return event_type == 'ai_dialogue'

File: src-tauri/src/ai_service/game_system/script_engine/events/ai_dialogue_event.rs
//! AI dialogue event — sets character and generates an AI reply via MessageGenerator.

use anyhow::{anyhow, Result};
use async_trait::async_trait;
use serde_json::Value;
use tauri::Manager;

use crate::ai_service::game_system::script_engine::events::{
    register_event, ScriptContext, ScriptEvent,
};
use crate::ai_service::game_system::script_engine::utils::script_function;
use crate::ai_service::message_system::generator::{GeneratorDeps, MessageGenerator};
use crate::ai_service::types::{LineAttributeExt, LineBase};
use crate::db::entities::line::LineAttribute;
use crate::utils::prompt::PromptRole;
use crate::AppState;

pub struct AIDialogueEvent {
    character: String,
    prompt: Option<String>,
}

impl AIDialogueEvent {
    fn from_event_data(data: &Value) -> Self {
        Self {
            character: data
                .get("character")
                .and_then(|v| v.as_str())
                .unwrap_or("MAIN")
                .to_string(),
            prompt: data
                .get("prompt")
                .and_then(|v| v.as_str())
                .map(|s| s.to_string()),
        }
    }
}

#[async_trait]
impl ScriptEvent for AIDialogueEvent {
    async fn execute(&mut self, ctx: &mut ScriptContext<'_>) -> Result<Option<String>> {
        let script_status = ctx
            .game_status
            .lock()
            .await
            .script_status
            .clone()
            .ok_or_else(|| anyhow!("ScriptStatus 未设置"))?;

        let (role_id, _role_display_name) = {
            let mut gs = ctx.game_status.lock().await;
            let role = script_function::get_role(&mut *gs, ctx.db, &script_status, &self.character)
                .await?;
            let id = role.role_id.ok_or_else(|| anyhow!("角色 ID 未设置"))?;
            let dn = role.display_name.clone();
            (id, dn)
        };

        // Set as current character
        ctx.game_status.lock().await.current_role_id = Some(role_id);

        // Inject prompt as SYSTEM line if provided
        // TODO: 这里的 prompt 是暂时的，应该标记为临时 prompt，并且在代码逻辑中在AI回复后清除这部分提示词。
        if let Some(ref prompt) = self.prompt {
            let sys_line = LineBase {
                content: PromptRole::Plot.build_prompt(prompt),
                attribute: LineAttributeExt(LineAttribute::User),
                display_name: Some("旁白".to_string()),
                ..Default::default()
            };
            ctx.game_status
                .lock()
                .await
                .add_line(ctx.db, sys_line)
                .await?;
        }

        // Delegate AI response to MessageGenerator
        let state = ctx.app.state::<AppState>();
        let llm = match state.chat.llm.clone() {
            Some(llm) => llm,
            None => {
                tracing::warn!("[AIDialogueEvent] LLM 未配置，跳过 AI 对话");
                return Ok(None);
            }
        };

        let deps = GeneratorDeps {
            app: ctx.app.clone(),
            db: ctx.db.clone(),
            game_status: ctx.game_status.clone(),
            processor: state.chat.processor.clone(),
            translator: state.chat.translator.clone(),
            llm,
            concurrency: 1,
        };

        let generator = MessageGenerator::new(deps);
        generator.process_message(None).await?;

        Ok(None)
    }

    fn event_type() -> &'static str {
        "ai_dialogue"
    }
}

pub fn register() {
    register_event(AIDialogueEvent::event_type(), |data| {
        Box::new(AIDialogueEvent::from_event_data(&data))
    });
}
